New solution_4 for C#

stephentoub opened the pull request:
## Description

Adds a new C# solution.  This is based off of the C# solution_2, with one main difference. It uses a bool[] instead of a ulong[].  This will result in 8-bit flags instead of 1-bit, but it means we don't need to do math manually, can use IndexOf on the array to find the next unmarked slot, and can more easily slice and follow patterns friendly to avoiding bounds checking.

## Contributing requirements

* [X] I read the contribution guidelines in CONTRIBUTING.md.
* [X] I placed my solution in the correct solution folder.
* [X] I added a README.md with the right badge(s).
* [X] I added a Dockerfile that builds and runs my solution.
* [X] I selected `drag-race` as the target branch.
* [X] All code herein is licensed compatible with BSD-3.

GieltjE commented:
Might be more than we are guessing, we have been avoiding it like the plague for even simple applications for a long, long time.

rbergen commented:
Closing this. We have a [drag-race C# solution](https://github.com/PlummersSoftwareLLC/Primes/tree/drag-race/PrimeCSharp/solution_3) in place that incorporates Stopwatch, string interpolation and some other differences from Dave's solution.

stephentoub committed:
New solution_4 for C#

Based off of solution_2.  Main changes from that one:
- Switches to a bool[] rather than ulong[]; increases bit size from 1 to 8, but enables optimizations elsewhere
- Uses IndexOf(false) to find the next prime candidate
- Avoids bounds checking via a sliced span when marking non-primes
- Uses Stopwatch instead of DateTime.UtcNow
- Fixes use-after-return issue for ArrayPool array

stephentoub commented:
On my machine, with .NET 5:

#### solution_2

Passes: 6881, Time: 5.0000237, Avg: 0.000726642014242116, Limit: 1000000, Count: 78498, Valid: True

davepl;6881;5.00002;1;algorithm=base,faithful=yes,bits=1

#### solution_4

Passes: 7911, Time: 5.0002114, Avg: 0.0006320580710403237, Limit: 1000000, Count: 78498, Valid: True

stephentoub;7911;5.00021;1;algorithm=base,faithful=yes,bits=8

stephentoub commented:
Some folks also tried this out on different machines, e.g. on a Mac M1:

#### solution_2

Passes: 7515, Time: 5.0001094, Avg: 0.0006653505522288757, Limit: 1000000, Count: 78498, Valid: True
 
davepl;7515;5.00011;1;algorithm=base,faithful=yes,bits=1

#### solution_4

Passes: 19875, Time: 5.0000552, Avg: 0.00025157510440251574, Limit: 1000000, Count: 78498, Valid: True
 
stephentoub;19875;5.00006;1;algorithm=base,faithful=yes,bits=8

stephentoub commented:
Oops... I just realized this is davepl's fork.  Sorry about that... will contribute it to the main repo.
